hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+- Grab the length of this string
+- Get modulo of the length of the string and the length of the internal array
+- Use this modulo as the index
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+### !callout-info
+
+## Why Modulo in This Example?
+
+When arrays get created, they get created with a fixed size in memory. This means that they have a maximum length and _can't_ hold an infinite number of things (unless there is an operation which expands the array. Python's list method `append` manages this for us!). Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!
+
+### !end-callout
+
+## Collisions
+
+There is an enormous number of possible keys and a limited-size array. At some point, a hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.
+
+### Example Collision
+
+Let's use the same hash function from before:
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+Imagine a hash table that stores "name"-"job title" key-value pairs. The hash table's internal `hash_array` has a length of `15`. Imagine inserting the following pairs, which have the following indices after applying the above hash function:
+
+| Key (Name) | Value (Job Title)     | Index (`hash_code(key)`) |
+| ---------- | --------------------- | ------------------------ |
+| `"dan"`    | `"Lead Instructor"`   | `3`                      |
+| `"chris"`  | `"CS Fun Instructor"` | `5`                      |
+| `"dee"`    | `"Lead Instructor"`   | `3`                      |
+
+If we insert Dan first, and then eventually Dee, our hash function will attempt to add Dee at the index `3`, even though Dan already exists there.
+
+### Resolving Collisions
+
+There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements into the collection.
+
+### Example Resolution
+
+If the values are stored in an array themselves, we can end up with a hash which looks like this:
+
+![Bad hash function example](../assets/hash-tables_hash-tables_collision.png)
+
+In this example, we resolve collisions by creating an array at each index of the main array, then adding the key-value pair to the end of the list.
+
+### !callout-info
+
+## Effective Hash Functions
+
+What would happen if we inserted 600 items, and our hash function happened to insert all 600 items into the same key? If the hash function does not adequately spread values out over the length of the array, we'll have long chains of key-value pairs and an O(n) worst case runtime to find items.
+
+Effective hash functions attempt to spread all possible values over the entire data structure to avoid assigning multiple keys to the same index. Most libraries use well-proven hashing functions which normally prevent high numbers of collisions.
```

Comment:
to me, prevent sounds more like the function is actively stopping something from happening, when really it's more like it's designed to do its best to avoid a situation it can't control.
```suggestion
Effective hash functions attempt to spread all possible values over the entire data structure to avoid assigning multiple keys to the same index. Most libraries use well-proven hashing functions which normally avoid high numbers of collisions.
```